### ISSUE TITLE
What's the way to calculate the 'new_alpha'?

### ISSUE THREAD

**Issue by jkstyle2** [cersar/3D_detection#4]
First, Thanks for your amazing work done!
I wonder, 
In the code 

                new_alpha = float(line[3]) + np.pi / 2.
                if new_alpha < 0:
                    new_alpha = new_alpha + 2. * np.pi
                new_alpha = new_alpha - int(new_alpha / (2. * np.pi)) * (2. * np.pi)

Can you please explain how this calculation has been made?

**Comment by cersar** [comment]
Hi，this piece of code comes from the reference repo in README.There，line[3] in kitti label file is the observation angle of object.Refence to https://blog.csdn.net/sum_nap/article/details/84858911, i think the gap between theta_loc and alpha is pi/2.This is meaning of the first line in the code.The rest, I guess, is just to make sure the alpha is between 0 and pi*2.

**Comment by jkstyle2** [comment]
Thanks for your comment, but still it doesn't make any sense to get the correct local orientation from that way. I consider we should add 3/(2*np.pi) to alpha to get the local orientation.
How do you think? Please help me understand..

**Comment by cersar** [comment]
Sorry,I just can't grasp your idea of adding 3/(2*np.pi).Can you explain more detailed about that?Thanks

**Comment by jkstyle2** [comment]
Um.. it's very difficult to explain what my idea is.. it would be better if I could show you a drawing by uploading a file.

Let me try describing it in words.
As long as I know, alpha is the angle between "the direction of an object" and "the perpendicular direction of the ray from the object's center". So, the local angle can be made as 270[deg] + alpha[deg].
It can also be expressed as 360[deg] - (ray angle[deg] -  rotation_y[deg]) , which is from here(https://github.com/shashwat14/Multibin/blob/master/data_prep.py), in line 49.
I'm not very sure though if I'm understanding it in a right way. 
If you're thinking that your code is correct, can you please elaborate it? It's actually widely used in other similar projects.

**Comment by jkstyle2** [comment]
I just drew it. Please correct me if I'm wrong with it.
![kakaotalk_20190123_225715779](https://user-images.githubusercontent.com/34189274/51611369-73b4de00-1f62-11e9-9394-e98569438068.jpg)

**Comment by cersar** [comment]
Ha,ha,it's really difficult to explain that without a picture, and thank you for your detailed explanation.
According to you drawing,i think there is some thing contrary to the kitti official README description.In the README.txt of 3D object labels, it says that alpha is zero only, when this object is facing
along the X-axis of the camera coordinate system and located along the Z-axis of the camera.In your drawing, when this object is facing along the X-axis of the camera coordinate system and located along the Z-axis of the camera,alpha is 270(deg),not zero.
My formal explain is reference this picture:
![20181206160004125](https://user-images.githubusercontent.com/9484017/51616574-58979d80-1f65-11e9-8ab1-f34e2ba2f858.png)
Where,alpha is the angle between object's heading and the X-axis of the camera coordinate system when the object's center is rotated along the Z-axis of the camera.
Forgive me for my poor English:-)

**Comment by cersar** [comment]
You are right!I have rethinked just now,and find the origin code has mistaked theta local by refering to the negative direction of ray from the object's center.So it is 180(deg) shifted to your answer.In the detection.py,I manually fixed it by pi/2-theta_loc.I think there will be no need to fix in the detection.py,if I fix the compution of theta_loc.Thank you very much for your advice,and it is really helpful to the project!I will fix it soon.

**Comment by jkstyle2** [comment]
Thanks for your quick response!
Of course, I checked the kitti's README.txt and I also misunderstood like what you said. But I realized that the saying "facing along the X-axis" means not looking at the x-axis, but it's going along the x-axis(parallel) . Think about the picture you uploaded, you will see that alpha is going to zero when the direction of the object become parallel with x-axis on the z-axis.

**Comment by jkstyle2** [comment]
hmm?

**Comment by cersar** [comment]
Ha,ha,this is a bug in the origin code.Just ignore my former reply.Thank you!

**Comment by jkstyle2** [comment]
Thanks for your help too! It's a very helpful project.
One thing now I'm concerning is, did the original code really mean it to get a local orientation? 
In the original code, the new_alpha goes to the function 'compute_anchor()', and the anchors are returned. 
As I'm not exactly sure of what the function does, I'm getting more confused now. Even other people who continue to implement the original code such as [smallcorgi](https://github.com/smallcorgi/3D-Deepbox/blob/master/data_processing.py), [scutan90](https://github.com/scutan90/YOLO-3D-Box/blob/master/data_processing.py) and etc, they all use the same code.
Are you certain that the original code really made a mistake?

Thanks for your help! Let's figure it out together.

**Comment by cersar** [comment]
Hi,i have made some updates in this weekend.
About the new_alpha,i find that we should compute new_alpha by the equation -new_alpha+3*pi/2.There, using '-new_alpha' instead of 'new_alpha' because the kitti's alpha is the angle from" the direction of an object" to "the perpendicular direction of the ray from the object's center", whereas the local orientation equals to the angle from "the perpendicular direction of the ray from the object's center" to " the direction of an object" added by the angle from "the direction of the ray from the object's center" to "the perpendicular direction of the ray from the object's center" (270 deg).
The function 'compute_anchor()' split 360 deg to several small bins, and compute which bins the the local orientation angle is to fall into according the overlap rate.I'm not sure that the original code really made a mistake about new alpha, but according to the paper it is reasonable to use the real as the input to the function 'compute_anchor()'.
According to the fix of new_alpha, i have trained another model which has been updated in the README.Take it easy to download the new one.

**Comment by skhadem** [comment]
Hello, I am also working on this issue in my PyTorch implementation. I am pretty new to machine learning. If alpha and theta_l are just 270(deg) apart, why not train the net to just output alpha? Is there a reason with the multibin approach where it wouldn't work as well? The reason I say this is that in my algorithm to calculate the 3D constraints I use alpha to narrow down the possibilities, so I figured I should just try and regress alpha directly instead of theta_l. Is there any reason this wouldn't work as well? Thanks
